documentation/src/docs/wire-protocol.md

```diff
@@ -9,104 +9,102 @@ You may obtain a copy of the License at
+1.  Segment that was read: A length of 2 byte String, followed by that many bytes of Java's Modified UTF-8).
+2.  Offset that was read from: 8 byte Long.
+3.  Is at Tail: 1 bit Boolean.
+4.  Is at `EndOfSegment`: 1 bit.
+5.  Data: Binary (remaining length in the message).
 
-Appending
----------
+The client requests to read from a particular Stream at a particular Offset. It then receives one or more replies in the form of `SegmentRead` messages. These contain the data they requested (assuming it exists). The server decides on more or less data to give the client than it asked for, in as many replies as it sees fit.
 
-### Setup Append - Request
+# Appending
 
-1.  ConnectionId - UUID (16 bytes) Identifies this appender.
-2.  Segment to append to. - String (2 byte length, followed by that many bytes of Java's Modified UTF-8)
+## Setup Append - Request
 
-### Append Setup - Reply
+1.  `ConnectionId`: UUID (16 bytes) Identifies this appender.
+2.  Segment to append: A length of 2 byte String, followed by that many bytes of Java's Modified UTF-8.
 
-1.  Segment that can be appended to. - String (2 byte length, followed by that many bytes of Java's Modified UTF-8)
-2.  ConnectionId - UUID (16 bytes) Identifies the requesting appender.
-3.  ConnectionOffsetAckLevel - Long (8 bytes) What was the last offset received and stored on this segment for this connectionId (0 if new)
+## Append Setup - Reply
 
-### BeginAppendBlock - Request 
+1.  Segment that can be appended: A length of 2 byte String, followed by that many bytes of Java's Modified UTF-8.
+2.  `ConnectionId`: UUID (16 bytes) Identifies the requesting appender.
+3.  `ConnectionOffsetAckLevel`: 8 byte Long. (The last offset received and stored on this Stream Segment for this `ConnectionId` (0 if new)).
 
-Only valid after SetupAppend has already been done successfully.
+## AppendBlock - Request
 
-1.  ConnectionId - UUID (16 bytes)
-2.  ConnectionOffset - Long (8 bytes) Data written so far over this connection to this segment
-3.  Length of data before EndAppendBlock message - Integer (4 bytes) 
+1. Final `UUID writerId`: (16 bytes).
+2. Final `ByteBuf` data: This holds the contents of the block.
 
-### EndAppendBlock- Request
+## AppendBlockEnd - Request
 
-1.  ConnectionId - UUID (16 bytes)
-2.  ConnectionOffset - Long (8 bytes) Data written so far over this connection
-3.  Block Length - (4 Bytes) Total size of the block that was written. (Note this may more or less than the number of bytes between the BeginAppendBlock and this message)
+1. Final `UUID writerId`: (16 bytes).
+2. Final `sizeOfWholeEvents`: 4 byte Integer
+3. Final `ByteBuf` data: This holds the contents of the block.
+4. Final `numEvents`: 4 byte Integer.
+5. Final  `lastEventNumber`: 8 byte Long.
 
-### Event - Request
 
-Only valid inside of a block
+
+## Event - Request
+
+Only valid inside the block.
 
 1.  Data
 
-### Data Appended - Reply
+## Data Appended - Reply
```

Comment:
Move the Partial Event to here, as it is also only valid inside an append block, and only at the end.

Reply:
Edited